contracts/strategies/allocation/direct-grants/DirectGrantsStrategy.sol

```diff
@@ -0,0 +1,317 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.19;
+
+import {IRegistry} from "../../../core/IRegistry.sol";
+import {BaseStrategy} from "../../BaseStrategy.sol";
+import {Metadata} from "../../../core/libraries/Metadata.sol";
+
+import {IERC20} from "@openzeppelin/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/token/ERC20/utils/SafeERC20.sol";
+import {IAllo} from "../../../core/IAllo.sol";
+
+contract DirectGrantsStrategy is BaseStrategy {
+    /// ===============================
+    /// ========== Errors =============
+    /// ===============================
+
+    error RECIPIENT_ALREADY_ACCEPTED();
+    error NO_ACCEPTED_RECIPIENT();
+    error UNAUTHORIZED();
+    error INVALID_MILESTONE();
+    error MILESTONE_ALREADY_ACCEPTED();
+
+    /// ================================
+    /// ========== Storage =============
+    /// ================================
+
+    string public constant VERSION = "2.0.0";
+
+    struct Milestone {
+        uint256 amountPercentage;
+        Metadata metadata;
+        uint256 payoutTimestamp;
+        RecipientStatus milestoneStatus;
+    }
+
+    struct Recipient {
+        address payoutAddress;
+        RecipientStatus status;
+        uint256 totalAmount;
+    }
+
+    struct Allocation {
+        address token;
+        uint256 amount;
+        address recipientId;
+    }
+
+    IRegistry public registry;
+
+    // recipientId => Recipient
+    mapping(address => Recipient) public recipients;
+
+    Milestone[] public milestones;
+
+    bool public registryGating;
+    uint256 public upcomingMilestone;
```

Comment:
this would likely have to be mapping as well from recipientId to upcomingMilestone. We could explore merging this into the mapping above